libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,222 @@
+
+  /// Calls finish on the active span, its parent is set to be active
+  /// Primarily for internal use only, preferred to use ProgressSpan's Finish function
+  void FinishActiveSpan();
+
+  // For passing spans across process/host boundaries
+  // These functions are needed, but are implemented now for
+  // debugging purposes, they will be replaced
+  // Extract returns a nullptr on failure
+  virtual std::string Inject(const ProgressContext& ctx) = 0;
+  virtual std::unique_ptr<ProgressContext> Extract(
+      const std::string& carrier) = 0;
+
+  /// Get the current scope’s span to add tagging/logging without having
+  /// to pass around scopes/spans as parameters
+  /// If there is no active span, an unnamed root span of a new trace
+  /// is created (in this case the program is probably not using tracing)
+  virtual ProgressSpan& GetActiveSpan();
+  bool HasActiveSpan() { return active_span_ != nullptr; }
+  uint32_t GetHostID() { return host_id_; }
+  uint32_t GetNumHosts() { return num_hosts_; }
+
+  /// Close is called when a tracer is finished processing spans
+  /// This should be called to ensure any and all buffered spans are
+  /// flushed
+  virtual void Close() = 0;
+
+private:
+  static std::unique_ptr<ProgressTracer> tracer_;
+
+  std::shared_ptr<ProgressSpan> active_span_ = nullptr;
+  uint32_t host_id_;
+  uint32_t num_hosts_;
+  std::shared_ptr<ProgressSpan> default_active_span_ = nullptr;
+};
+
+class KATANA_EXPORT ProgressScope {
+  friend class ProgressTracer;
+
+  ProgressScope(std::shared_ptr<ProgressSpan> span) : span_(span) {}
+
+public:
+  ~ProgressScope();
+  ProgressScope(const ProgressScope&) = delete;
+  ProgressScope(ProgressScope&&) = delete;
+  ProgressScope& operator=(const ProgressScope&) = delete;
+  ProgressScope& operator=(ProgressScope&&) = delete;
+
+  /// Get the scope’s underlying span to add tagging/logging,
+  /// span relationships, and handle multiple active spans at a time
+  ProgressSpan& span() { return *span_; }
+
+  /// Closes the underlying ProgressSpan if the ProgressScope was created
+  /// with the flag finish_on_close=true
+  /// Note that this will only close the underlying ProgressSpan when all
+  /// of its active children ProgressSpans have been finished
+  /// This will be called by RAII if it has not already been called
+  void Close();
+
+private:
+  std::shared_ptr<ProgressSpan> span_;
+};
+
+class KATANA_EXPORT ProgressContext {
+public:
+  virtual ~ProgressContext() = default;
+  virtual std::unique_ptr<ProgressContext> Clone() const noexcept = 0;
+  virtual std::string GetTraceID() const noexcept;
+  virtual std::string GetSpanID() const noexcept;
+};
+
+class KATANA_EXPORT ProgressSpan {
+  virtual void Close() = 0;
+
+protected:
+  ProgressSpan(std::shared_ptr<ProgressSpan> parent) : parent_(parent) {}
+
+public:
+  /// If Finish has not already been called for the ProgressSpan, it's
+  /// destructor must do so.
+  virtual ~ProgressSpan() = default;
```

Comment:
probably delete move and copy member functions.

Reply:
To be clear for these you want me to do something like:
  ProgressTracer(const ProgressTracer&) = delete;
  ProgressTracer(ProgressTracer&&) = delete;
  ProgressTracer& operator=(const ProgressTracer&) = delete;
  ProgressTracer& operator=(ProgressTracer&&) = delete;
  
and then reset them to default in the child classes like:
  JSONTracer(const JSONTracer&) = default;
  JSONTracer(JSONTracer&&) = default;
  JSONTracer& operator=(const JSONTracer&) = default;
  JSONTracer& operator=(JSONTracer&&) = default;

Reply:
clang doesn't like this so I am going to remove it

Reply:
I'm not sure why you'd need to reset them in the child classes. `= default` (confusingly) means "compiler, synthesize the function if you can". In the case where your parent class is not movable/copyable, the "default" implementation for a child class is not viable:

https://godbolt.org/z/Kj9xrz9n6

and it would be clearer to omit them altogether, which is sort of the point of deleting the special member functions in the base class to begin with. If we intend to access objects through their base class, we probably shouldn't let objects be copied otherwise objects will be sliced.

Is it the case that we need JSONTracer to be copyable? If so, we'd have to define the special member functions explicitly, though I would ask under what circumstances we need a Tracer to be copyable, as that seems to go against the pointer based API we have now.